### PR TITLE
An initial implementation of the phase noise channel

### PR DESCRIPTION
**Pull request by arsalan-motamedi**
**Context:**
Adds the phase noise with a Fock representation.

**Description of the Change:**
Added a class under `Channel`. Please note that if we wanted to actually write a Fock array for this channel, it would be extremely heavy (I ran into memory issues). Therefore, we implement contractions with `PhaseNoise` in a particular way, using a custum rrshift method. I believe a similar logic was used in lab before.

**Benefits:**
Enables us to apply phase noise.

**Possible Drawbacks:**
Could be slow.

### CODE REVIEW

File: mrmustard/lab_dev/states/base.py [hunk 1]
```diff
@@ -50,7 +51,7 @@
 )
 
 from ..circuit_components import CircuitComponent
-from ..circuit_components_utils import BtoPS
```
Comment (apchytr) [inline mrmustard/lab_dev/states/base.py:53]: Can you revert the changes to states?

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+class PhaseNoise(Channel):
+    r"""
+    The Phase noise channel.
+
+    This class represents the application of a random phase. The distributiuon of the phase
+    is assumed to be a Gaussian with mean zero, and standard deviation `sigma`.
+
+    Args:
+        modes: The modes the channel is applied to
+        sigma: the standard deviation of the random Gaussian noise (could be provided as a list, if the channel acts on more than one mode)
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:41]:
```suggestion
        sigma: The standard deviation of the random Gaussian noise.```

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    ..details::
+        The Fock representation is connected to the Fourier coefficients of the distribution.
+
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:45]:
```suggestion
```

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    """
+
+    short_name = "P~"
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:48]: Maybe `PNoise`?

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    # randomized : bool
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:50]:
```suggestion
```

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    def __init__(
+        self,
+        modes: Sequence[int],
+        phase_stdev: float | Sequence[float],
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:54]: Should this be `sigma` or should the docstring be `phase_stdev`?

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        phase_stdev_trainable: bool = False,
+        phase_stdev_bounds: tuple[float | None, float | None] = (0.0, None),
+    ):
+        super().__init__(name="PhN")
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:58]:
```suggestion
        super().__init__(name="PhaseNoise")
```

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        self._add_parameter(
+            make_parameter(phase_stdev_trainable, phase_stdev, "phase_stdev", phase_stdev_bounds)
+        )
+        self._representation = self.from_ansatz(
+            modes_in=modes, modes_out=modes, ansatz=None
+        ).representation
+
+    def __custom_rrshift__(self, other):
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:66]: Can you add type hints?

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        r"""
+        Custom rrshift
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:68]: Can you either remove the docstring or have something more descriptive?

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+                    array[array_index] *= np.exp(
+                        -0.5
+                        * (idx[mode] - idx[other.n_modes + mode]) ** 2
+                        * self.phase_stdev.value**2
+                    )
+            return CircuitComponent(
+                Representation(ArrayAnsatz(array, False), dm_state.wires), self.name
+            )
+        else:
+            raise ValueError("The PhaseNoise object can only be applied to a DM or a Ket.")
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:91]:
```suggestion
            raise ValueError("PhaseNoise can only be applied to a DM or a Ket.")
```

File: mrmustard/physics/fock_utils.py [hunk 11]
```diff
@@ -36,6 +36,7 @@
     wigner_to_bargmann_U,
 )
 from mrmustard.utils.typing import ComplexTensor, Matrix, Scalar, Tensor, Vector, Batch
+import itertools
```
Comment (apchytr) [inline mrmustard/physics/fock_utils.py:39]:
```suggestion
```

File: tests/test_lab_dev/test_transformations/test_phasenoise.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+from mrmustard.lab_dev.transformations import Dgate, PhaseNoise
+
+
+class TestPhaseNoise:
+    r"""
+    Tests for the ``PhaseNoise`` class.
+    """
+
+    def test_init(self):
+        "Tests the GaussRandNoise initialization."
```
Comment (apchytr) [inline tests/test_lab_dev/test_transformations/test_phasenoise.py:29]:
```suggestion
        "Tests the PhaseNoise initialization."
```

**Review by apchytr — COMMENTED** [review]
Looks great thanks!

File: mrmustard/lab_dev/transformations/phasenoise.py (hunk 10 above)
Comment (ziofil) [reply mrmustard/lab_dev/transformations/phasenoise.py:91]: Maybe there's an issue with this. If what started as a Ket or DM contracts with other stuff it will no longer be a Ket or a DM, but it's still correct to apply the PhaseNoise to those wires.

**Review by ziofil — COMMENTED** [review]
Great work Arsalan! But see my comment about that potential issue

File: mrmustard/lab_dev/states/base.py (hunk 1 above)
Comment (arsalan-motamedi) [reply mrmustard/lab_dev/states/base.py:53]: Yes, I did

File: mrmustard/lab_dev/transformations/phasenoise.py (hunk 10 above)
Comment (apchytr) [reply mrmustard/lab_dev/transformations/phasenoise.py:91]: Doesn't the rshift of Ket and DM guarantee that anytime a state is contracted with a transformation it returns a state?

File: mrmustard/lab_dev/transformations/phasenoise.py (hunk 4 above)
Comment (apchytr) [reply mrmustard/lab_dev/transformations/phasenoise.py:48]: Agreed but is `P~` clear enough as a short name?

File: mrmustard/lab_dev/states/base.py (hunk 1 above)
Comment (apchytr) [reply mrmustard/lab_dev/states/base.py:53]: Weird theres still a diff? Did you do `git checkout develop ./mrmustard/lab_dev/states`?

File: mrmustard/lab_dev/transformations/phasenoise.py (hunk 10 above)
Comment (arsalan-motamedi) [reply mrmustard/lab_dev/transformations/phasenoise.py:91]: I guess it is fixed now?

File: mrmustard/lab_dev/transformations/phasenoise.py (hunk 4 above)
Comment (ziofil) [reply mrmustard/lab_dev/transformations/phasenoise.py:47]: yes. let's make it consistent, so that other noise channels have the tilde

File: mrmustard/lab_dev/transformations/phasenoise.py (hunk 10 above)
Comment (ziofil) [reply mrmustard/lab_dev/transformations/phasenoise.py]: What I'm saying is that if the object that initially is a Ket or DM contracts with something else before it contracts with PhaseNoise, it can be that the resulting object isn't a Ket or a DM and the contraction with PhaseNoise would fail, despite being the correct thing to do next.

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+            for count, _ in enumerate(np.nditer(array)):
+                idx = np.zeros(len(array.shape))
+                temp = count
+                for l in range(len(idx)):
+                    idx[-1 - l] = temp % array.shape[-1 - l]
+                    temp = temp // array.shape[-1 - l]
+                array_index = tuple(idx.astype(int))
+                array[array_index] *= np.exp(
+                    -0.5 * (idx[mode] - idx[other.n_modes + mode]) ** 2 * self.phase_stdev.value**2
+                )
```
Comment (ziofil) [inline mrmustard/lab_dev/transformations/phasenoise.py:88]:
```suggestion
        array = math.asnumpy(other.fock_array())
        mode_indices = np.indices(array.shape)
        for mode in self.modes:
                phase_factors = np.exp(
                -0.5 * (mode_indices[mode] - mode_indices[n_modes + mode])**2 * self.phase_stdev.value**2
        )
        array *= phase_factors
```
I believe this does the same and it should be more efficient.

File: tests/test_lab_dev/test_transformations/test_phasenoise.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        assert ch.name == "PhaseNoise"
+        assert ch.phase_stdev.value == 0.2
+        assert ch.modes == [0, 1]
+        assert ch.ansatz == None
+
+    def test_application(self):
+        "Tests application of PhaseNoise on Ket and DM"
+        psi = Ket.random([0, 1]) >> Dgate([0], 0.5, 0.5) >> PhaseNoise([0], 0.2)
+        assert isinstance(psi, DM)
+        assert psi.purity < 1
```
Comment (ziofil) [inline tests/test_lab_dev/test_transformations/test_phasenoise.py:40]: it would be great to hardcode a test where we know a couple of actual values

File: mrmustard/lab_dev/transformations/phasenoise.py (hunk 10 above)
Comment (arsalan-motamedi) [reply mrmustard/lab_dev/transformations/phasenoise.py]: Well, now it works on those cases Filippo, no?

File: mrmustard/lab_dev/states/base.py [hunk 15]
```diff
@@ -42,6 +42,7 @@
 )
 from mrmustard.physics.fock_utils import quadrature_distribution
 from mrmustard.physics.wigner import wigner_discretized
+from ..circuit_components_utils import BtoPS
```
Comment (apchytr) [inline mrmustard/lab_dev/states/base.py:45]: Can you revert this change?

File: mrmustard/lab_dev/states/dm.py [hunk 16]
```diff
@@ -32,10 +32,11 @@
 from mrmustard.physics.representations import Representation
 from mrmustard.physics.wires import Wires
 from mrmustard.utils.typing import ComplexMatrix, ComplexVector, ComplexTensor, RealVector
+from ..circuit_components_utils import BtoQ, TraceOut
```
Comment (apchytr) [inline mrmustard/lab_dev/states/dm.py:35]: Can you revert this also?

File: mrmustard/lab_dev/states/ket.py [hunk 17]
```diff
@@ -31,6 +31,7 @@
 from mrmustard.physics.gaussian import purity
 from mrmustard.physics.representations import Representation
 from mrmustard.physics.wires import Wires
+from ..circuit_components_utils import BtoQ, TraceOut
```
Comment (apchytr) [inline mrmustard/lab_dev/states/ket.py:34]: This too please 😃

File: mrmustard/lab_dev/states/ket.py [hunk 18]
```diff
@@ -371,6 +373,10 @@ def __rshift__(self, other: CircuitComponent | Scalar) -> CircuitComponent | Bat
         with those of a ``DM`` or of a ``Ket``. Returns a ``CircuitComponent`` in general,
         and a (batched) scalar if there are no wires left, for convenience.
         """
+        from ..transformations.phasenoise import PhaseNoise
+
+        if isinstance(other, PhaseNoise):
+            return self.dm() >> other
```
Comment (apchytr) [inline mrmustard/lab_dev/states/ket.py:379]: Please remove

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+from __future__ import annotations
+from typing import Sequence
+from mrmustard import math
+from mrmustard.physics.representations import Representation
+from mrmustard.physics.ansatz.array_ansatz import ArrayAnsatz
+from mrmustard.lab_dev.circuit_components import CircuitComponent
+from .base import Channel
+from ..utils import make_parameter
+import numpy as np
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:27]: Can you move this above line 21

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+"""
+The class representing a Phase noise channel.
+"""
+
+from __future__ import annotations
+from typing import Sequence
+from mrmustard import math
+from mrmustard.physics.representations import Representation
+from mrmustard.physics.ansatz.array_ansatz import ArrayAnsatz
+from mrmustard.lab_dev.circuit_components import CircuitComponent
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:24]: Can you make these imports (line 22-24) alphabetical?

File: mrmustard/lab_dev/transformations/phasenoise.py [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+class PhaseNoise(Channel):
+    r"""
+    The Phase noise channel.
+
+    This class represents the application of a random phase. The distributiuon of the phase
+    is assumed to be a Gaussian with mean zero, and standard deviation `phase_stdev`.
+
+    Args:
+        modes: The modes the channel is applied to
+        phase_stdev: The standard deviation of the random Gaussian noise.```
```
Comment (apchytr) [inline mrmustard/lab_dev/transformations/phasenoise.py:41]:
```suggestion
        phase_stdev: The standard deviation of the random Gaussian noise.
```

File: tests/test_lab_dev/test_transformations/test_phasenoise.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+    Tests for the ``PhaseNoise`` class.
+    """
+
+    def test_init(self):
+        "Tests the PhaseNoise initialization."
+        ch = PhaseNoise([0, 1], 0.2)
+        assert ch.name == "PhaseNoise"
+        assert ch.phase_stdev.value == 0.2
+        assert ch.modes == [0, 1]
+        assert ch.ansatz == None
```
Comment (apchytr) [inline tests/test_lab_dev/test_transformations/test_phasenoise.py:39]:
```suggestion
        assert ch.ansatz is None
```

File: mrmustard/lab/gates.py [hunk 23]
```diff
@@ -933,7 +933,7 @@ class Amplifier(Channel):
     """
 
     is_gaussian = True
-    short_name = "Amp"
+    short_name = "Amp~"
```
Comment (apchytr) [inline mrmustard/lab/gates.py:936]: Just out of curiosity why these changes?

**Review by apchytr — COMMENTED** [review]
Thanks Arsalan!

File: mrmustard/lab/gates.py (hunk 23 above)
Comment (arsalan-motamedi) [reply mrmustard/lab/gates.py:936]: Filippo suggested adding tilde to all noise channels, so that P~ is meaningful enough.

**Review by ziofil — APPROVED** [review]
looks good!

**Review by apchytr — APPROVED** [review]
